Remove wrap_dim from codegen layer.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32738 Remove wrap_dim from codegen layer.**

This is to simplify the codegen layer, with the goal of making it simple enough to just check in.

Differential Revision: [D19610927](https://our.internmc.facebook.com/intern/diff/D19610927)

gchanan committed:
Remove wrap_dim from codegen layer.

This is to simplify the codegen layer, with the goal of making it simple enough to just check in.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ab660194:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ab6601940038beab909944958b8415ac80f03f7c).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4364431) pytorch_linux_xenial_py3_6_gcc5_4_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92270196) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92270196)) :snowflake:
<details>
<summary>
<code>Jan 29 00:19:30 RuntimeError: quantization scale should be > 0</code>
</summary>

```
Jan 29 00:19:30   File "test_quantization.py", line 730, in test_quantized_rnn 
Jan 29 00:19:30     y, (h, c) = cell_dq(x, (h, c)) 
Jan 29 00:19:30   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 540, in __call__ 
Jan 29 00:19:30     result = self.forward(*input, **kwargs) 
Jan 29 00:19:30   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 373, in forward 
Jan 29 00:19:30     return self.forward_tensor(input, hx) 
Jan 29 00:19:30   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 334, in forward_tensor 
Jan 29 00:19:30     input, hx, batch_sizes, max_batch_size, sorted_indices) 
Jan 29 00:19:30   File "/opt/conda/lib/python3.6/site-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 315, in forward_impl 
Jan 29 00:19:30     self.batch_first, dtype=self.dtype, use_dynamic=True) 
Jan 29 00:19:30 RuntimeError: quantization scale should be > 0 
Jan 29 00:19:30  
Jan 29 00:19:30 ---------------------------------------------------------------------- 
Jan 29 00:19:30 Ran 36 tests in 54.261s 
Jan 29 00:19:30  
Jan 29 00:19:30 FAILED (errors=1, skipped=1) 
Jan 29 00:19:30  
Jan 29 00:19:30 Generating XML reports... 
Jan 29 00:19:30 Traceback (most recent call last): 
Jan 29 00:19:30   File "test/run_test.py", line 457, in <module> 
Jan 29 00:19:30     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

gchanan committed:
Update on "Remove wrap_dim from codegen layer."


This is to simplify the codegen layer, with the goal of making it simple enough to just check in.

Differential Revision: [D19610927](https://our.internmc.facebook.com/intern/diff/D19610927)

[ghstack-poisoned]

gchanan committed:
Update on "Remove wrap_dim from codegen layer."


This is to simplify the codegen layer, with the goal of making it simple enough to just check in.

Differential Revision: [D19610927](https://our.internmc.facebook.com/intern/diff/D19610927)

[ghstack-poisoned]

bhosmer commented on aten/src/THC/THCTensorScatterGather.cu:
```diff
@@ -3,6 +3,7 @@
 #include <THC/THCAtomics.cuh>
 #include <THC/THCApply.cuh>
 #include <c10/macros/Macros.h>
+#include <ATen/WrapDimUtils.h>
```
Ditto

bhosmer reviewed (approved):
So nice... thanks for doing this!

gchanan commented on aten/src/THC/THCTensorIndex.cu:
```diff
@@ -17,6 +17,7 @@
 #include <thrust/sort.h>
 #include <algorithm> // for std::min
 #include <c10/macros/Macros.h>
+#include <ATen/WrapDimUtils.h>
```
It's just a way of including the file once instead of # of dtypes times.

The way the TH macros work is:
aten/src/TH(C)/TH(C)MyFile.ext includes
aten/src/TH(C)/generic/TH(C)MyFile.ext 1 time for each dtype, setting what e.g. `scalar_t` means.

So in this case it's the generic version that actually uses the functions here, I'm just including it in the non-generic version to simplify the include logic.

gchanan committed:
Update on "Remove wrap_dim from codegen layer."


This is to simplify the codegen layer, with the goal of making it simple enough to just check in.

Differential Revision: [D19610927](https://our.internmc.facebook.com/intern/diff/D19610927)

[ghstack-poisoned]
